File: Antlr/src/main/scala/io/github/chaosunity/casc/parsing/expression/IfExpression.scala
package io.github.chaosunity.casc.parsing.expression

import io.github.chaosunity.casc.exceptions.TypeUnmatchedException
import io.github.chaosunity.casc.parsing.`type`.Type

class IfExpression(val condition: Expression,
                   val trueExpression: Expression,
                   val falseExpression: Expression) extends Expression {
    override def `type`: Type = if (trueExpression.`type` == falseExpression.`type`) trueExpression.`type` else
        throw new TypeUnmatchedException(trueExpression.`type`, falseExpression.`type`)

    override def negative: Boolean = false
}

File: src/main/kotlin/org/casclang/casc/parsing/node/expression/IfExpression.kt
package org.casclang.casc.parsing.node.expression

import org.casclang.casc.parsing.type.Type

data class IfExpression(
    val condition: Expression<*>,
    val trueExpression: Expression<*>,
    val falseExpression: Expression<*>
) : Expression<IfExpression> {
    override val type: Type
        get() = if (trueExpression.type == falseExpression.type) trueExpression.type else
            throw RuntimeException("Type unmatched for type '${trueExpression.type.internalName}' and type '${falseExpression.type.internalName}'")
}